tests/data/test_dataset.py

```diff
@@ -0,0 +1,55 @@
+from numbers import Number
+
+import numpy as np
+
+from pennylane.data import AttributeInfo, Dataset, DatasetScalar
+
+
+class TestDataset:
+    def test_init(self):
+        """Test that initializing a Dataset with keyword arguments
+        creates the expected attributes.
+        """
+        ds = Dataset(
+            description="test",
+            x=DatasetScalar(1, AttributeInfo(doc="A variable")),
+            y=np.array([1, 2, 3]),
+            z="abc",
+        )
+
+        assert ds.description == "test"
+        assert ds.x == 1
+        assert (ds.y == np.array([1, 2, 3])).all()
+        assert ds.z == "abc"
+
+    def test_setattr(self):
+        """Test that __setattrr__ succesfully sets new and existing attributes."""
+        ds = Dataset(description="test", x=1)
+
+        ds.x = 2.0
```

Comment:
Does the type of `x` change? Can we assign it to a string?

Reply:
Good question. Attributes that or not in `fields` are replaced completely when assigned to. So yes, the type of `x` will change. For attributes that are in `fields`, the declared type of the attribute will be used, and the info will be preserved.